pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+underscore is already heavily used as an 'ignore value' marker in iterable
+unpacking.  Since the wildcard pattern ``_`` never binds, this use of the
+underscore does not interfere with other uses such as inside the REPL or
+internationalisation packages.
+
+Finally note that the underscore is as a wildcard pattern in *every*
+programming language with pattern matching that we could find.  Keeping
+in mind that many users of Python also work with other programming
+languages, have prior experience when learning Python, or moving on to
+other languages after having learnt Python, we find that such well
+established standards are important and relevant with respect to
+readability and learnability.  Moreover, concerns that this wildcard
+means that a regular name received special treatment are not strong
+enough to introduce syntax that would make Python special.
+
+
+.. _literal_pattern:
+
+Literal Patterns
+~~~~~~~~~~~~~~~~
+
+Literal patterns are a convenient way for imposing constraints on the
+value of a subject, rather than its type or structure.  Literal patterns
+even allow you to emulate a switch statement using pattern matching.  On
+the flipside, if you think of patterns as building on parameters and
+assignment targets, literal patterns are a novel addition (i.e. you would
+not write, e.g., ``(2, a, b) = c`` in iterable unpacking).
+
+Originally, literal patterns came from the idea of expressing unstructured
+singleton objects such as ``None``: instead of requiring that a subject has
+type ``NoneType``, it makes much more sense to directly write ``None``.
+More generally, literal patterns could also be seen as syntactic sugar for
+guards.  Rather than ``case x if x == 0:``, you can simply write ``case 0:``.
+
+Generally, the subject is compared to a literal pattern by means of standard
+equality (``x == y`` in Python syntax).  Consequently, the literal patterns
+``1.0`` and ``1`` match exactly the same set of objects, i.e. ``case 1.0:``
+and ``case 1:`` are fully interchangable.  In principle, ``True`` would also
+match the same set of objects because ``True == 1`` holds.  However, we
+believe that many users would be surprised findings that ``case True:``
+matched the object ``1.0``, resulting in some subtle bugs and convoluted
+work arounds.  We therefore adopted the rule that the three singleton
+objects ``None``, ``False`` and ``True`` match by identity (``x is y`` in
+Python syntax) rather than equality.  Hence, ``case True:`` will match only
+``True`` and nothing else.  Note that ``case 1:`` would still match ``True``,
+though, because the literal pattern ``1`` works by equality and not identity.
+
+Early ideas to induce a hierarchy on numbers so that ``case 1.0`` would
+match both the integer ``1`` and the floating point number ``1.0``, whereas
+``case 1:`` would only match the integer ``1`` very eventually dropped in
+favour of the simpler and consistent rule based on equality.
```

Comment:
Maybe give this more color? Implementing this correctly would be quite expensive, because we'd have to use things like `isinstance(x, numbers.Integral)`, which is expensive (as it's a protocol, not just an ABC). There's not really much precedent of such behavior elsewhere in Python -- `x == 1` matches 1, True, 1.0, even 1.0+0j. And you can write `case int(1)` if you want to insist on an int, or `case i := numbers.Integral() if i == 1` if you want to use the numeric tower.